### PR TITLE
AniDB wrapper API fetches creators and creates artists

### PR DESCRIPTION
**Pull request by mrsalixor**
See issue #382.

### CODE REVIEW

File: mangaki/mangaki/utils/anidb.py [hunk 1]
```diff
@@ -183,6 +181,33 @@ def get_or_update_work(self,
...
+                # AniDB marks Studio as such a creator's type
+                studio, s_created = Studio.objects.get_or_create(title=creator)
+
+            if staff_id is not None:
+                creators.append({
+                    "role_id": staff_map[staff_id],
+                    "name": creator,
+                    "anidb_creator_id": creator_id
+                })
+                artist = Artist.objects.get_or_create(name="Shinbou Akiyuki")
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/anidb.py]: Why do you fetch this particular artist?

File: mangaki/mangaki/utils/anidb.py [hunk 2]
```diff
@@ -183,6 +181,33 @@ def get_or_update_work(self,
             if title_type == 'synonym':
                 synonyms[lang] = title
 
+        # Handling of staff
+        creators = []
+        staff_map = dict(Role.objects.filter(slug__in=['author', 'director', 'composer']).values_list('slug', 'pk'))
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/anidb.py]: Arguably, you don't have to bother restricting the queryset to these slugs only.
You could just fetch all `Role`s objects.

Also, let's put a `# FIXME: cache this query`, like `lang_cache` because the contents will rarely change during the lifetime of a deployment.

File: mangaki/mangaki/utils/anidb.py [hunk 3]
```diff
@@ -183,6 +181,33 @@ def get_or_update_work(self,
...
+
+            if creator_type == 'Direction':
+                staff_id = 'director'
+            elif creator_type == 'Music':
+                staff_id = 'composer'
+            elif creator_type == 'Original Work' or creator_type == 'Story Composition':
+                staff_id = 'author'
+            elif creator_type == 'Animation Work' or creator_type == 'Work':
+                # AniDB marks Studio as such a creator's type
+                studio, s_created = Studio.objects.get_or_create(title=creator)
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/anidb.py]: We should definitely add a new field called `anidb_aid` to `Studio` I guess.
Let's do this in another PR.

File: mangaki/mangaki/tests/test_anidb.py [hunk 4]
```diff
@@ -4,22 +4,27 @@
...
             return f.read()
 
     def setUp(self):
         # FIXME: The defaults for editor and studio in Work requires those to
         # exist, or else foreign key constraints fail.
         Editor.objects.create(pk=1)
-        Studio.objects.create(pk=1)
-        self.anidb = client
+        # Studio.objects.create(pk=1)
+        self.anidb = AniDB('mangakihttp', 1)
```
Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_anidb.py]: Why did you have to change the client?

File: mangaki/mangaki/tests/test_anidb.py [hunk 5]
```diff
@@ -4,22 +4,27 @@
...
+        with open(os.path.join(settings.TEST_DATA_DIR, filename), 'r', encoding='utf-8') as f:
             return f.read()
 
     def setUp(self):
         # FIXME: The defaults for editor and studio in Work requires those to
         # exist, or else foreign key constraints fail.
         Editor.objects.create(pk=1)
-        Studio.objects.create(pk=1)
-        self.anidb = client
+        # Studio.objects.create(pk=1)
```
Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_anidb.py]: It's pretty dangerous, currently, the `Work` model uses a `default` of `Studio(id=1)` when no studios are specified.
But I guess it's okay since AniDB give us `studios` :D.

File: mangaki/mangaki/tests/test_anidb.py [hunk 6]
```diff
@@ -4,22 +4,27 @@
...
 class AniDBTest(TestCase):
     @staticmethod
+    def create_anime(**kwargs):
+        anime = Category.objects.get(slug='anime')
+        return Work.objects.create(category=anime, **kwargs)
+
+    @staticmethod
     def read_fixture(filename):
-        with open(os.path.join(settings.TEST_DATA_DIR, filename), 'r') as f:
+        with open(os.path.join(settings.TEST_DATA_DIR, filename), 'r', encoding='utf-8') as f:
```
Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_anidb.py]: Good job, as most of us are Linux / Mac developers, we have `utf-8` as default locale encoding. We wouldn't be able to catch this kind of bug.

File: mangaki/mangaki/tests/test_anidb.py [hunk 7]
```diff
@@ -47,5 +52,19 @@ def test_anidb_get(self):
...
         )
         anime = self.anidb.get_or_update_work(11606)
-        self.assertNotEqual(anime.title, '')
-        self.assertNotEqual(anime.worktitle_set.count(), 0)
+
+        self.assertEqual(anime.title, 'Sangatsu no Lion')
+        self.assertEqual(anime.nb_episodes, 22)
+        self.assertEqual(anime.studio.title, 'Shaft')
+
+        author_ids = [q.artist_id for q in Staff.objects.filter(work_id=anime.pk, role_id=3)]
```
Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_anidb.py]: This could be rewritten in an idiomatic Django as: `Staff.objects.filter(work_id=anime.pk, role_id=3).values_list('artist_id', flat=True)` (which will return a list of IDs).

Also, rather than to rely on numeric value such as ID, you could filter with `role__slug='author'`.

[About the `field__something`, it's called a lookup.](https://docs.djangoproject.com/fr/1.11/ref/models/querysets/#field-lookups)
You can either use a field of the foreign key considered, `role__slug` or `role__name` for example.
Or you could use integrated lookups to have different results `role__slug__contains='a'`.

File: mangaki/mangaki/tests/test_anidb.py [hunk 8]
```diff
@@ -47,5 +52,19 @@ def test_anidb_get(self):
...
+        author_names = [Artist.objects.get(pk=at_id).name for at_id in author_ids]
```
Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_anidb.py]: In order to get a list of items knowing their ID, in idiomatic Django, you could write:

`Artist.objects.in_bulk(author_ids)` and you will get a dictionary of `{ id: Artist object }`.

File: mangaki/mangaki/tests/test_anidb.py [hunk 9]
```diff
@@ -47,5 +52,19 @@ def test_anidb_get(self):
...
+        self.assertEqual(author_names, ['Umino Chika'])
```
Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_anidb.py]: Asserting between lists should be done using [`assertCountEqual`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual) rather than `assertEqual`.

For example, `self.assertEqual(['A', 'B'], ['B', 'A'])` would fail, using `assertCountEqual` on these lists would succeed.

Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_anidb.py]: ## Another way to get `author_names`

Finally, getting the author names of a given work could be done using the related fields.

First, `Staff` has a relationship of type Many to One with `Work` (it's marked by the `ForeignKey('work')` which is on `Staff` model).

Now, as a result, `Work` get a special "invisible" field called `staff_set` which returns a queryset of `Staff` which all respects the condition `Staff.work = Work considered`.

Example: `Work.objects.get(id=1).staff_set.all()` would return the list of staff of the work with `id=1`.

Knowing this, you could fetch all staff members using `anime.staff_set` and then partition the list by `role` (a sort of `group_by`).

Finally, you could access to the name using `staff.artist.name`.

The question is: which one is faster? which one should we use? Let me know your opinion on this.

Comment (mrsalixor) [inline mangaki/mangaki/tests/test_anidb.py]: I looked up your recommandations to improve queries to the database in `test_anidb.py`. I tried a few solutions scrolling down your comments, and thought it would be interesting to write a bit about it ! But yeah, the last recommandation is indeed the best, as expected since it reduces redundancy of queries and makes use of relationships between models.

*Note : woops, this text might be longer than I had anticipated when starting to write. Sorry !*


#### My first solution basically made too many queries if we are facing a big staff :
- 1 query to get the list of IDs of staff corresponding to the work id and the role id (so 3 queries in total)
- 1 query per ID of staff, in order to retrieve the name of that staff

Therefore, we basically add up to `(3 + 3*N)*A` queries where `N` is the size of the staff and `A` the number of animes.

```py
author_ids = [q.artist_id for q in Staff.objects.filter(work_id=anime.pk, role_id=3)]
author_names = [Artist.objects.get(pk=at_id).name for at_id in author_ids]

composer_ids = [q.artist_id for q in Staff.objects.filter(work_id=anime.pk, role_id=4)]
composer_names = [Artist.objects.get(pk=cp_id).name for cp_id in composer_ids]

director_ids = [q.artist_id for q in Staff.objects.filter(work_id=anime.pk, role_id=5)]
director_names = [Artist.objects.get(pk=dt_id).name for dt_id in director_ids]
```

#### I tried out a second solution very similar to the first one bug using Django's lookups (in this case, `pk__in`) as you mentioned in a previous comment. This makes it more easily readable by using role slugs and slightly reduces the count of queries if we are facing a big staff :
- Still 1 query to get the list of IDs of staff corresponding to the work id and the role id (so 3 queries total)
- But only 1 query to retrieve the names of the staff given their ID (so 3 queries total)

This basically gets us to `6A` queries where `A` the number of animes.

```py
author_ids = Staff.objects.filter(work_id=anime.pk, role__slug='author').values_list('artist_id', flat=True)
author_names = Artist.objects.filter(pk__in=author_ids).values_list('name', flat=True)

composer_ids = Staff.objects.filter(work_id=anime.pk, role__slug='composer').values_list('artist_id', flat=True)
composer_names = Artist.objects.filter(pk__in=composer_ids).values_list('name', flat=True)

director_ids = Staff.objects.filter(work_id=anime.pk, role__slug='director').values_list('artist_id', flat=True)
director_names = Artist.objects.filter(pk__in=director_ids).values_list('name', flat=True)
```

#### Now, digging up Django's documentation a bit more, I found out that I could bring those 2 queries to a single one per role using relationships between models. It basically means 3 queries per work id, so `3A` queries in total. Moreover, queries are way faster since I now make real use of relationships, which I didn't do until now :

```py
author_names = Staff.objects.filter(work_id=anime.pk, role__slug='author').values_list('artist__name', flat=True)
composer_names = Staff.objects.filter(work_id=anime.pk, role__slug='composer').values_list('artist__name', flat=True)
director_names = Staff.objects.filter(work_id=anime.pk, role__slug='director').values_list('artist__name', flat=True)
```

#### However, I'm still doing the same query for each role. In order to further reduce computation time, it is indeed better to retrieve the list of staff beforehand and then only pick out the artist names corresponding to a specific role thanks to a `filter`. This brings us down to 1 query per work id and just filtering afterwards :

```py
staff = Work.objects.get(pk=anime.pk).staff_set.all() # Only called once per work !

author_names = staff.filter(role__slug='author').values_list('artist__name', flat=True)
composer_names = staff.filter(role__slug='composer').values_list('artist__name', flat=True)
director_names = staff.filter(role__slug='director').values_list('artist__name', flat=True)
```

I don't see a possibility to further reduce computation time and the number of queries at the moment (given there is one ...). I tried to loop these in order to see how it affects computation time at a bigger scale. Here were the results for 5000 iterations :

- **First solution :** 22.339s (first run) ; 23.364s (second run) ; 22.824s (third run)
- **Second solution :** 21.259s (first run) ; 21.381s (second run) ; 21.452s (third run)
- **Third solution :** 15.063s (first run) ; 14.917s (second run) ; 15.338s (third run)
- **Fourth solution :** 13.246s (first run) ; 13.277s (second run) ; 13.252s (third run)

Comment (RaitoBezarius) [inline mangaki/mangaki/tests/test_anidb.py]: It's an absolutely awesome write-up!

I think you're most likely right for your conclusion, the last way I see to fetch the names is to filter before `all()` by `role__slug__in=('author', 'composer', 'director')` and add `role__slug` in your `values_list` so that you can differentiate the `roles` and build the different lists using `itertools.groupby` or something similar.

How did you perform your measurements of timing?
For my part, I got 6.87 ms per loop (best of 3) over 5000 iterations using [`timeit`](https://docs.python.org/3/library/timeit.html) which is often the right way to perform such measurements.
